tempesta_fw/http_parser.c

```diff
@@ -457,6 +457,134 @@ parse_int_hex(unsigned char *data, size_t len, unsigned long *acc)
 	return CSTR_POSTPONE;
 }
 
+/**
+ * Add unconditionaly hop-by-hop headers to list of hop-by-hop headers names.
+ * Do not init raw headers table until we found that Connection header contains
+ * atleast one raw header.
+ */
+static void
+__hbh_parser_init(TfwHttpHbhHdrs *hbh_hdrs, bool req)
+{
+	/*
+	 * Connection header is hop-by-hop header.
+	 *
+	 * Server header isn't defined as hop-by-hop by the RFC, but we
+	 * don't show protected server to world. User-agent and Server headers
+	 * have the same id in tfw_http_hdr_t, so Server header must be marked
+	 * only for responses.
+	 */
+	hbh_hdrs->spec |= 0x1 << TFW_HTTP_HDR_CONNECTION;
+	if (!req)
+		hbh_hdrs->spec |= 0x1 << TFW_HTTP_HDR_SERVER;
+}
+
+#define MARK_SPEC_HBH(hid)						\
+	if ((msg->parser.hbh_parser.spec & (0x1 << hid))		\
+	    && !TFW_STR_EMPTY(&msg->parser.hdr))			\
+		msg->parser.hdr.flags |= TFW_STR_HBH_HDR;
+
+/**
+ * Mark raaw header @hdr as hop-by-hop if its name was listed in Connection
+ * header
+ */
+static void
+mark_raw_hbh(TfwHttpMsg *hm, TfwStr *hdr)
+{
+	TfwHttpHdrTbl *ht = hm->parser.hbh_parser.raw;
+	unsigned int i;
+
+	if (!ht)
+		return;
+
+	for (i = 0; i < ht->off; ++i) {
```

Comment:
1. RFC does not have any restrictions where headers listed in Connection header must be placed. So both variants are ok: before or after Connection header. More over i tried to set up custom hop-by-hop headers in ngingx and apache, both allow setting custom headers before and after the header. Here example from RFC 6455 with hop-by-hop header before Connection header:
```
GET /chat HTTP/1.1
Host: server.example.com
Upgrade: websocket
Connection: Upgrade
Sec-WebSocket-Key: dGhlIHNhbXBsZSBub25jZQ==
Origin: http://example.com
Sec-WebSocket-Protocol: chat, superchat
Sec-WebSocket-Version: 13
```

2. Table you point to stores names for only raw headers. I agree, it could be confusing, but i used data structure which suit me perfectly. In situation you show user-agent and every other spec header listed in connection header (except keep-alive) will not be removed

Obsoleted now, in new revision this array is statically allocated